src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java

```diff
@@ -0,0 +1,118 @@
+package se.liu.ida.hefquin.engine.datastructures.impl;
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+
+import java.util.*;
+
+public class SolutionMappingsHashTableBasedOnTwoVars extends SolutionMappingsHashTable{
+    protected final Map<Node, List<Node>> mapNN = new HashMap<>();
+    protected final Map<Node, List<SolutionMapping>> mapNS = new HashMap<>();
```

Comment:
I don't think that this is going to work. What would be the keys in `mapNS`? These keys should not be independent of the corresponding value for the other join variable. For instance, assume we first insert a solution mapping with var1=X and var2=Z, and after that we insert a solution mapping with var1=Y and var2=Z. Hence, both of these solution mappings have the same value for the second join variable. Then, when searching for join partners with var1=X and var2=Z, there should be an efficient way of retrieving only the relevant candidates with var1=X and var2=Z.

To fix this, the internal data structure here should be a nested map:
```
protected final Map<Node, Map<Node, List<SolutionMapping>>> map = new HashMap<>();
```
Then, the keys in the outer map (i.e., in `map`) are the RDF terms for the first join variable and the corresponding values would be further maps. The keys in each of these inner maps are the RDF terms for the second join variable, but only the ones that belong to the corresponding binding for the first variable.

Reply:
It makes sense. I have updated the data structure to a nested map.